sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc

```diff
@@ -0,0 +1,458 @@
+std::vector<sycl::event> depEvents = /* some dependencies */;
+sycl::event e = q.parallel_for(sycl::nd_range<1>{1024, 16}, depEvents,
+  [=](sycl::nd_item<1> it) {
+  result[it.get_global_id()] += 1;
+});
+e.wait();
+----
+
+
+==== Proposed syntax
+
+[source,c++]
+----
+using syclex = sycl::ext::oneapi::experimental;
+
+syclex::submit(q, [&](sycl::handler& h) {
+  sycl::accessor result { buf, h, sycl::write_only, sycl::no_init };
+  syclex::parallel_for(h, 1024, [=](sycl::id<1> idx) {
+    result[idx] = idx;
+  });
+});
+
+std::vector<sycl::event> depEvents = /* some dependencies */;
+sycl::event e;
+syclex::nd_launch(q, sycl::nd_range<1>{1024, 16}, [=](sycl::nd_item<1> it) {
+  result[it.get_global_id()] += 1;
+}, depEvents, &e);
+e.wait();
+----
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_ENQUEUE_FUNCTIONS` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Command group submission
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename ExecutorType, typename CommandGroupFunc>
+void submit(ExecutorType ex, CommandGroupFunc cgf, const std::vector<event>& depEvents = {}, event* outEvent = nullptr);
+
+}
+----
+
+[source,c++]
+----
+template <typename ExecutorType, typename CommandGroupFunc>
+void submit(ExecutorType ex, CommandGroupFunc cgf, const std::vector<event>& depEvents = {}, event* outEvent = nullptr);
+----
+_Constraints_: Available only if `ExecutorType` is `sycl::queue`.
+
+_Effects_: Submit a command group function object to be scheduled for execution
+on `ex`. The submitted command will not execute until each of the events in
+`depEvents` is complete. An optional event representing this command can be
+returned via `outEvent`.
+
+
+=== Single tasks
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename KernelName, typename ExecutorType, typename KernelType
+void single_task(ExecutorType ex, const KernelType& kernelFunc);
+
+template <typename KernelName, typename ExecutorType, typename KernelType
+void single_task(ExecutorType ex, const KernelType& kernelFunc, const std::vector<event>& depEvents = {}, event* outEvent = nullptr);
+
+}
+----
+
+[source,c++]
+----
+template <typename KernelName, typename ExecutorType, typename KernelType
+void single_task(ExecutorType ex, const KernelType& kernelFunc);
+----
+_Constraints_: Available only if `ExecutorType` is `sycl::handler`.
```

Comment:
Should this also allow `ExecutorType` to be `sycl::queue`?  I think we need this for your example in Overview to compile.

Reply:
My intent here was as follows.

`single_task(handler, [](){})` resolves to the first overload. The user can't add any events, because the `single_task` overload with two extra parameters isn't available.

`single_task(queue, [](){})` resolves to the second overload. The user is free to add events if they want to (because the second overload has two extra parameters for that purpose), but if those parameters are missing then they're defaulted.

I don't think we can allow both overloads to accept `sycl::queue`, because the call without events would then be ambiguous.  I think if we wanted `sycl::queue` to be accepted in both cases, we'd actually have to define three overloads:

1. `single_task(ExecutorType, const KernelType&)`
2. `single_task(ExecutorType, const KernelType&, const std::vector<sycl::event>&)`
3. `single_task(ExecutorType, const KernelType&, const std::vector<sycl::event>&, sycl::event*)`

Reply:
Oh, I got confused by this sentence because I missed the word "not":

> _Constraints:_ Available only if `ExecutorType` is not `sycl::handler`.

Personally, I think it would be clearer if you described it in the positive way:

> _Constraints:_ Available only if `ExecutorType` is `sycl::queue`.

If we allow other executor types in the future, we can adjust the wording at that time.

Reply:
Makes sense.  Changed in a360654d89d39c7815eb82035a937d8dcde15c06.